lib/async_hooks.js

```diff
@@ -259,8 +262,15 @@ function validateAsyncId(asyncId, type) {
 
 // Embedder API //
 
+const undestroyedAsyncIds = new Set();
```

Comment:
Are `Set`s a thing in `lib/`, or is that too much of a problem for perf?

Reply:
It’s perfectly fine.

Reply:
What’s the reason for this, though? Avoiding double cleanup when the implementor emits `destroy` manually? Can you add a comment?

Reply:
Exactly. Will do.

Reply:
If possible, I would like to avoid yet another global state and yet another special emit function :)

Is it not possible to read a property from the `handle` object in `MakeCallback`? Then we could just have another symbol property on in the `handle` object that indicates if `emitDestroy` was called manually.

Forcing the user to either always call `emitDestroy()` or otherwise never call `emitDestroy()`, depending on the optional flag is also an option.

Reply:
Which `handle` are you referencing? I can't find anything like that in any MakeCallback fn.

I guess using the flag as an exclusive choice might be the easiest thing to do here, just wanted to cover all the use-cases while we're at it, but it might make things more complicated.

Reply:
@Sebmaster Sorry, I meant `WeakCallback`.

> just wanted to cover all the use-cases while we're at it

What are the other use-cases.

Reply:
Ahh I see what you mean. That should be possible; I think v8 doesn't clear up the object before the weak callback gets invoked. I'll implement that.

> What are the other use-cases?

I was thinking about resources which allow explicit disposal, like imagine a connection where you could call `conn.close()` or just let it fall out of scope. I think the expectation there would be for the destroy event to emitted at the time of the close call rather than at GC time (necessitating a way to prevent a double call).

Reply:
After thinking about it some, that would not allow us to set the symbol on AsyncResource if the asyncId is destroyed by the sensitive API (since we don't have an AsyncResource reference available there).

Does there currently not exist a map with all async ids where we could store state in at all? I thought until now that `async_id_fields` did so we could shove an additional field in there, but just realized that that's just global state management after looking at it.

Reply:
Don't worry about the sensitive API. We will be deprecating that anyway and I think it is perfectly fine to require the user call `emitDestroy` manually if they insist on using a low-level API.

> Does there currently not exist a map with all async ids where we could store state in at all?

No such map exists.

Reply:
Unfortunately, after testing it, it seems like v8 frees the object before the weakcallback is called, so we can't attach props there either 😞

Reply:
Here we go, found a solution by introducing another object into it for a level of indirection. 🎉 

That good with you?